secrets/store.go

```diff
@@ -27,65 +38,73 @@ type Store struct {
 //
 // Context should come with a timeout otherwise this might block forever, i.e.
 // if the path never becomes available.
-func NewStore(ctx context.Context, path string, logger log.Wrapper) (*Store, error) {
-	parser := func(r io.Reader) (interface{}, error) {
-		secrets, err := NewSecrets(r)
-		if err != nil {
-			return nil, err
-		}
-		return secrets, nil
+func NewStore(ctx context.Context, path string, logger log.Wrapper, middlewares ...SecretMiddleware) (*Store, error) {
+	store := &Store{}
+	if len(middlewares) > 0 {
+		store.secretHandler(middlewares...)
 	}
-	result, err := filewatcher.New(ctx, path, parser, logger)
+
+	result, err := filewatcher.New(ctx, path, store.parser, logger)
 	if err != nil {
 		return nil, err
 	}
-	return &Store{
-		watcher: result,
-	}, nil
+
+	store.watcher = result
+	return store, nil
 }
 
-// GetSimpleSecret fetches a simple secret or error if the key is not present.
-func (s *Store) GetSimpleSecret(path string) (SimpleSecret, error) {
-	var secret SimpleSecret
-	data := s.watcher.Get()
-	secrets, ok := data.(*Secrets)
-	if !ok {
-		return secret, fmt.Errorf("unexpected type %T", data)
+func (s *Store) parser(r io.Reader) (interface{}, error) {
+	secrets, err := NewSecrets(r)
+	if err != nil {
+		return nil, err
 	}
-	secret, ok = secrets.simpleSecrets[path]
-	if !ok {
-		return secret, ErrorSecretNotFound(path)
+
+	if s.secretHandlerFunc != nil {
+		s.secretHandlerFunc(secrets)
 	}
-	return secret, nil
+
+	return secrets, nil
 }
 
-// GetVersionedSecret fetches a versioned secret or error if the key is not present.
-func (s *Store) GetVersionedSecret(path string) (VersionedSecret, error) {
-	var secret VersionedSecret
+// SecretHandler creates the middleware chain.
+func (s *Store) secretHandler(middlewares ...SecretMiddleware) {
+	if s.secretHandlerFunc == nil {
+		s.secretHandlerFunc = nopSecretHandlerFunc
+	}
+
+	for _, m := range middlewares {
+		s.secretHandlerFunc = m(s.secretHandlerFunc)
+	}
+}
+
+// GetSimpleSecret loads secrets from watcher, and fetches a simple secret from secrets
+func (s *Store) GetSimpleSecret(path string) (SimpleSecret, error) {
```

Comment:
Should we make a breaking change so that `Store` no longer provides `Get*Secret` functions directly, but return the `Secret` instead and user can use `Get*Secret` from `Secret`?

Reply:
I'm very hesitated about it. Ideally we should be introducing v2 if there's breaking change, but is there anything using this at the moment? If not, it's not a breaking change, if there is, how much work are involved to cascade the change?

Reply:
We are at v0. You can make breaking changes on v0 (but should be careful about them).

Reply:
Also cc @konradreiche

Reply:
From a different perspective, `Store` is the primary entry point to secrets, it might be cleaner to keep these methods so that in most client code we can do
```go
        sec, err := Store.Get*Secret()
```

instead of
```go
        secrets, err := Store.GetSecrets()
        if err != nil {
                ...
        }
        sec, err := secrets.Get*Secret()
```

Reply:
Is there a reason `GetSecrets` could return an error? If not this would just be `store.GetSecrets().Get*Secret()`.

Reply:
function chaining in Go is not very common and often leads to some problems. The original code does type assert so there's an error to check
```
	secrets, ok := s.watcher.Get().(*Secrets)
```

Reply:
That type assertion would never fail: https://godoc.org/github.com/reddit/baseplate.go/filewatcher#Result.Get

Reply:
oh... right. haha.

OK gonna make one last argument here:
`Store` stores a `*Secrets`, every time there's a file change, `Store` creates a new `*Secrets` and put it into filewatch data. I think have a `GetSecrets() *Secrets` is an error prone API, user may thought it's ok to keep using returned `*Secrets` since it's a pointer to the value, but in fact they should call `GetSecrets()` every time. And if they do use the same `*Secrets`, there's no error to report the problem.

We need to document this well, or maybe have `GetSecrets` return a struct and call it out the object may _expire_ when there's a file change, either way it could still lead to problem unknowingly.

Reply:
Yea I agree that we need to document this well, but this is similar to the whole "don't cache any secrets and always get them before using" document we already have.

Reply:
If we are happy with it I can make the change. To me the current way is less likely to lead to problems. It's a little redundant, but considering the functions attached to `Secrets` are pretty much only used when there's a callback need, I'm ok with it.

Reply:
I'll wait for Konrad to chime in :)